octodns/provider/plan.py

```diff
@@ -67,19 +77,32 @@ def raise_if_unsafe(self):
            len(self.existing.records) >= self.MIN_EXISTING_RECORDS:
 
             existing_record_count = len(self.existing.records)
-            update_pcent = self.change_counts['Update'] / existing_record_count
-            delete_pcent = self.change_counts['Delete'] / existing_record_count
+            if existing_record_count > 0:
+                update_pcent = (self.change_counts['Update'] /
+                                existing_record_count)
+                delete_pcent = (self.change_counts['Delete'] /
+                                existing_record_count)
+            else:
+                update_pcent = 0
+                delete_pcent = 0
 
             if update_pcent > self.update_pcent_threshold:
-                raise UnsafePlan('Too many updates', update_pcent * 100,
-                                 self.update_pcent_threshold * 100,
-                                 self.change_counts['Update'],
-                                 existing_record_count)
+                raise TooMuchChange('Too many updates', update_pcent * 100,
+                                    self.update_pcent_threshold * 100,
+                                    self.change_counts['Update'],
+                                    existing_record_count)
             if delete_pcent > self.delete_pcent_threshold:
-                raise UnsafePlan('Too many deletes', delete_pcent * 100,
-                                 self.delete_pcent_threshold * 100,
-                                 self.change_counts['Delete'],
-                                 existing_record_count)
+                raise TooMuchChange('Too many deletes', delete_pcent * 100,
+                                    self.delete_pcent_threshold * 100,
+                                    self.change_counts['Delete'],
+                                    existing_record_count)
+
+        # If we have any changes of the root NS record for the zone it's a huge
+        # deal and force should always be required for extra care
+        if self.exists and [True for c in self.changes
+                            if c.record and c.record._type == 'NS' and
+                            c.record.name == '']:
+            raise RootNsChange()
```

Comment:
Looks like any change to root NS is marked unsafe, not just deletes. Is that true and intentional? It will be difficult to setup DNS CI pipelines to detect root NS changes and pass `--force` dynamically, more so if someone wants to stay within update/delete thresholds but allow root NS updates.

Reply:
That is intentional. It seemed safest given the scenario where a user has always had "bad" root NS records in their config and they update to a newer octoDNS and provider which now supports root NS management. Their next `--doit` will push the wrong root NS records and unless they're paying really close attention they might not notice. I.e. thinking that all they did was update versions and make no record changes they might not look closely enough to see them.

It doesn't have to be that way, but it seemed safer to do so.

`--force` is akin to saying a human as taken notice that something noteworthy is happening here and has Okayed the changes. They're not intended as things that will never happen, just a hey something might be up here. 

As such I would expect anyone running things as part of a CI/CD pipeline would generally have to handle `--force` already. I always ran things w/chatops so it was easy enough to add the `--force` flag. If i were doing it as part of a CI/CD setup where the apply happens I'd probably have the force driven off of a special sign-off comment in the PR or some such. Specifics are hard to make up without all the details of the setup.

Reply:
Ok I think that makes sense. Trying to think what happens during a new zone creation. Would that also require a `--force`?

Reply:
Should only fire when the zone is already exists, e.g. the `exists` part of the clause. So not on zone creates, but it would if you're creating a root NS record in a zone that doesn't have one (PowerDNS is the only thing I know of where that's possible.)

Reply:
Hmm interesting, should be ok then.